baseapp/abci.go

```diff
@@ -230,18 +249,31 @@ func (app *BaseApp) Commit() (res abci.ResponseCommit) {
 	// empty/reset the deliver state
 	app.deliverState = nil
 
-	defer func() {
-		if app.haltHeight > 0 && uint64(header.Height) == app.haltHeight {
-			app.logger.Info("halting node per configuration", "height", app.haltHeight)
-			os.Exit(0)
-		}
-	}()
-
 	return abci.ResponseCommit{
 		Data: commitID.Hash,
 	}
 }
 
+func (app *BaseApp) halt() {
+	app.logger.Info("halting node per configuration", "height", app.haltHeight, "time", app.haltTime)
+
+	p, err := os.FindProcess(os.Getpid())
+	if err == nil {
+		// attempt cascading signals in case SIGINT fails (os dependent)
+		sigIntErr := p.Signal(syscall.SIGINT)
+		sigTermErr := p.Signal(syscall.SIGTERM)
```

Comment:
Trying to understand the logic here, you're sending yourself a SIGINT to attempt to finishing the defers? And we're sending both signals assuming if the first one fails the second one will get processed?
What if sigint succeeds?

Another question, should we print out the errs here, or they're all useless info here?

Reply:
>Trying to understand the logic here, you're sending yourself a SIGINT to attempt to finishing the defers?

No, you're sending a SIGINT/SIGTERM to the current (main) process. This will be captured by [`TrapSignal`](https://github.com/cosmos/cosmos-sdk/blob/v0.37.0/server/start.go#L148) which in turn will invoke any cleanup and defer logic.

> And we're sending both signals assuming if the first one fails the second one will get processed? What if sigint succeeds?

Not exactly, we're sending both to be compliant with both unix and windows. If one or the other fails, it's benign as long as at least one succeeds which is the happy case. If so, we'll return and halt gracefully. The unhappy path is if both fail and we have to `os.Exit(0)`.